number/src/macros.rs

```diff
@@ -0,0 +1,256 @@
+macro_rules! powdr_field {
+    ($name:ident, $ark_type:ty) => {
+        use crate::{traits::FieldElementTrait, AbstractNumberType, DegreeType};
+        use ark_ff::{BigInteger, Field, PrimeField};
+        use std::fmt;
+        use std::ops::*;
+
+        #[derive(Clone, Copy, PartialEq, Eq, Debug, Default, PartialOrd, Ord, Hash)]
+        pub struct $name {
+            value: $ark_type,
+        }
+
+        impl FieldElementTrait for $name {
+            fn to_degree(&self) -> DegreeType {
+                let degree = self.to_integer();
+                let mut limbs = degree.iter_u64_digits();
+                let res = (&mut limbs).next().unwrap_or_default();
+                assert!(limbs.next().is_none());
+                res
+            }
+
+            fn to_integer(&self) -> AbstractNumberType {
+                self.value.into_bigint().into()
+            }
+
+            fn modulus() -> AbstractNumberType {
+                <$ark_type>::MODULUS.into()
+            }
+
+            fn pow(self, exponent: AbstractNumberType) -> Self {
+                Self {
+                    value: self.value.pow([exponent.try_into().unwrap()]),
+                }
+            }
+
+            fn integer_div(self, other: Self) -> Self {
+                (self.to_integer() / other.to_integer()).into()
+            }
+
+            fn to_bytes_le(&self) -> Vec<u8> {
+                self.value.into_bigint().to_bytes_le()
+            }
+        }
+
+        impl num_traits::Zero for $name {
+            fn zero() -> Self {
+                Self::from(0)
+            }
+
+            fn is_zero(&self) -> bool {
+                self.value.is_zero()
+            }
+        }
+
+        impl<V: Into<$ark_type>> From<V> for $name {
+            fn from(value: V) -> Self {
+                Self {
+                    value: value.into(),
+                }
+            }
+        }
+
+        // Add
+
+        impl std::ops::Add for $name {
+            type Output = $name;
+
+            fn add(self, rhs: Self) -> Self::Output {
+                $name {
+                    value: self.value + rhs.value,
+                }
+            }
+        }
+
+        impl<'a> std::ops::Add<&'a $name> for $name {
+            type Output = $name;
+
+            fn add(self, rhs: &'a $name) -> Self::Output {
+                $name {
+                    value: self.value + rhs.value,
+                }
+            }
+        }
+
+        impl<'a> std::ops::Add for &'a $name {
+            type Output = $name;
+
+            fn add(self, rhs: Self) -> Self::Output {
+                $name {
+                    value: self.value + rhs.value,
+                }
+            }
+        }
+
+        impl<'a> std::ops::Add<$name> for &'a $name {
```

Comment:
Are the implementations for the references not automatic?

Reply:
I removed these as it's not needed now that the underlying integer is Copy